expression/builtin_time.go

```diff
@@ -1918,139 +1932,541 @@ func (b *builtinExtractDurationSig) evalInt(row []types.Datum) (int64, bool, err
+		sig = &builtinDateAddIntIntSig{baseTimeBuiltinFunc{bf}}
+	case dateEvalTp == tpDatetime && intervalEvalTp == tpString:
+		sig = &builtinDateAddDatetimeStringSig{baseTimeBuiltinFunc{bf}}
+	case dateEvalTp == tpDatetime && intervalEvalTp == tpInt:
+		sig = &builtinDateAddDatetimeIntSig{baseTimeBuiltinFunc{bf}}
+	}
 	return sig.setSelf(sig), nil
 }
 
-type builtinDateArithSig struct {
-	baseBuiltinFunc
-
-	op ast.DateArithType
+type builtinDateAddStringStringSig struct {
+	baseTimeBuiltinFunc
+}
+type builtinDateAddStringIntSig struct {
+	baseTimeBuiltinFunc
+}
+type builtinDateAddIntStringSig struct {
+	baseTimeBuiltinFunc
+}
+type builtinDateAddIntIntSig struct {
+	baseTimeBuiltinFunc
+}
+type builtinDateAddDatetimeStringSig struct {
+	baseTimeBuiltinFunc
+}
+type builtinDateAddDatetimeIntSig struct {
+	baseTimeBuiltinFunc
 }
 
-func (b *builtinDateArithSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func dateArithmeticalGetStringDate(ctx context.Context, args []Expression, row []types.Datum, unit string) (types.Time, bool, error) {
+	dateStr, isNull, err := args[0].EvalString(row, ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return types.Time{}, true, errors.Trace(err)
 	}
-	// args[0] -> Date
-	// args[1] -> Interval Value
-	// args[2] -> Interval Unit
-	// health check for date and interval
-	if args[0].IsNull() || args[1].IsNull() {
-		return
+
+	dateTp := mysql.TypeDate
+	if !types.IsDateFormat(dateStr) || types.IsClockUnit(unit) {
+		dateTp = mysql.TypeDatetime
 	}
-	nodeDate := args[0]
-	nodeIntervalValue := args[1]
-	nodeIntervalUnit := args[2].GetString()
-	if nodeIntervalValue.IsNull() {
-		return
+
+	date, err := types.ParseTime(dateStr, dateTp, types.MaxFsp)
+	if err != nil {
+		return types.Time{}, true, errors.Trace(handleInvalidTimeError(ctx, err))
 	}
-	// parse date
-	fieldType := mysql.TypeDate
-	var resultField *types.FieldType
-	switch nodeDate.Kind() {
-	case types.KindMysqlTime:
-		x := nodeDate.GetMysqlTime()
-		if (x.Type == mysql.TypeDatetime) || (x.Type == mysql.TypeTimestamp) {
-			fieldType = mysql.TypeDatetime
-		}
-	case types.KindString:
-		x := nodeDate.GetString()
-		if !types.IsDateFormat(x) {
-			fieldType = mysql.TypeDatetime
-		}
-	case types.KindInt64:
-		x := nodeDate.GetInt64()
-		if t, err1 := types.ParseTimeFromInt64(x); err1 == nil {
-			if (t.Type == mysql.TypeDatetime) || (t.Type == mysql.TypeTimestamp) {
-				fieldType = mysql.TypeDatetime
-			}
+	return date, false, nil
+}
+
+func dateArithmeticalGetIntDate(ctx context.Context, args []Expression, row []types.Datum, unit string) (types.Time, bool, error) {
+	dateInt, isNull, err := args[0].EvalInt(row, ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return types.Time{}, true, errors.Trace(err)
+	}
+
+	dateTp := mysql.TypeDate
+	if t, err := types.ParseTimeFromInt64(dateInt); err == nil {
```

Comment:
types.ParseTimeFromInt64 is for what? if err == nil, why not return t directlly?